rfc/rbac-ng.md

```diff
@@ -0,0 +1,154 @@
+It does not implement any form of access controls to its APIs beyond the authentication requirements of any reverse proxy in front of it such as Gate.
+Normally, access controls would be enforced via Fiat, though this approach has not scaled well and duplicates permission storage with a polling mechanism.
+Fiat is designed to collect resource metadata from other Spinnaker services and provide permission evaluation in return.
+The primary information Fiat provides access to are external role information for users along with a cache of application permissions which are otherwise stored by Front50.
+
+Kayenta is intended to work as a standalone application, however, and makes a good case study for an improved model for _role-based access controls_ (aka _RBAC_) across Spinnaker.
+To do this, the concept of _role policies_ will be introduced as an abstraction to link together roles and permissions to define the security policy of a resource such as an account.
+[Open Policy Agent](https://www.openpolicyagent.org) (OPA) is a flexible policy engine commonly used for handling authorization, compliance, and other policy-related concerns by evaluating expressions on documents (structured schemaless data).
+Fiat provides a natural facade for externalizing security policy enforcement with an engine like OPA.
+Starting with Kayenta, OPA needs to be informed of applications, metric store accounts, external roles, and role policies.
+Metric store accounts and role policies will need backing storage to allow for runtime management.
+Account credential management APIs should be added that work with these new access controls and allow permissions to be defined and managed on accounts while being configured.
+This will make Kayenta more multi-tenant friendly and provide guidance on how to further enhance RBAC in Spinnaker.
+
+### Goals and Non-Goals
+
+Goals:
+- Add support for role policies in Fiat with Gate APIs
+- Add API for storing and managing account definitions in Kayenta to integrate with user secrets
+- Add Open Policy Agent as a storage and permission evaluator backend for Fiat
+- Add support for role policies to Kayenta metric store accounts
+- Update Front50 to push application permission changes to Fiat
+
+Non Goals:
+- Provide APIs for managing security policies directly via Rego or other potential policy backends
+
+## Motivation and Rationale
+
+Kayenta APIs are all provided fairly openly assuming an admin has configured it with usable account credentials.
+Account credentials can be configured for metric stores and object stores, the latter of which can also be used for configuration stores.
+Object stores are used for storing and retrieving metric query results which are each assigned random UUIDs upon creation.
+Configurations are encouraged to be shared and are viewable by all applications by default; as such, these are the only object stores that list objects from the store.
```

Comment:
Could use clarification on
> these are the only object stores that list objects from the store

Reply:
I mean the object keys are only enumerated for config stores, not object stores in general.

Reply:
OK, I wonder if other folks get tripped up on the language here, or if it's just me.  I'm not coming up with a better alternative at the moment though...

Reply:
Those are the three account types in Kayenta, though that itself is an odd design choice. What I'm getting at here is that it seems as though the metric store accounts are the most important to provide access controls on, but the object stores might also benefit depending on how sensitive your metrics are.